### PR TITLE
common/gateio/stream: add thread-safe counter and overide default GenerateMessageID with connection specific implementation

### PR DESCRIPTION
**Pull request by shazbert**
# PR Description

* GenerateMessageID does not ensure ~~atomicity~~ collision resistance, this update adds a basic counter for routines that require low cost message matching signatures. 
* Gateio is converted by using bespoke ID generation in connection setup to override default `GenerateMessageID`
* NOTE: The overhead of `default` GenerateMessageID is quite high

Fixes # (issue)

* Signature collisions in outbound requests for Gateio. 

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [ ] go test ./... -race
- [ ] golangci-lint run
- [ ] Test X

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: common/common.go [hunk 1]
```diff
@@ -671,3 +672,19 @@ func SortStrings[S ~[]E, E fmt.Stringer](x S) S {
...
+
+// Counter is a thread-safe counter.
+type Counter struct {
+	n int64 // privatised so you can't use counter as a value type
+}
+
+// IncrementAndGet returns the next count after incrementing.
+func (c *Counter) IncrementAndGet() int64 {
+	newID := atomic.AddInt64(&c.n, 1)
+	// Handle overflow by resetting the counter to 1 if it becomes negative
```
Comment (gloriousCode) [inline common/common.go:684]: Had thoughts of suggesting uint64 instead, but there are many unknowns and I doubt anyone is going to subscribe to WIF more 9,223,372,036,854,775,807 times on GateIO in a single session

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Only one request, this is fine 🐶

**Review by gloriousCode — APPROVED** [review]
tACK!

# 🌞 
```
[DEBUG] | WEBSOCKET | 16/08/2024 13:37:42 | GateIO websocket connection: sending message [{"time":1723779462,"id":1,"channel":"spot.tickers","event":"subscribe","payload":["BTC_USDT"]}]
[DEBUG] | WEBSOCKET | 16/08/2024 13:37:42 | GateIO websocket connection: message received: {"time":1723779462,"time_ms":1723779462323,"id":1,"conn_id":"1b73153490cc89f8","trace_id":"daa79b26d14a54b172e11012230c1c78","channel":"spot.tickers","event":"subscribe","payload":["BTC_USDT"],"result":{"status":"success"},"requestId":"daa79b26d14a54b172e11012230c1c78"}
[DEBUG] | WEBSOCKET | 16/08/2024 13:37:42 | GateIO websocket connection: sending message [{"time":1723779462,"id":2,"channel":"spot.candlesticks","event":"subscribe","payload":["5m","BTC_USDT"]}]
[DEBUG] | WEBSOCKET | 16/08/2024 13:37:42 | GateIO websocket connection: message received: {"time":1723779462,"time_ms":1723779462670,"id":2,"conn_id":"1b73153490cc89f8","trace_id":"daa79b26d14a54b172e11012230c1c78","channel":"spot.candlesticks","event":"subscribe","payload":["5m","BTC_USDT"],"result":{"status":"success"},"requestId":"daa79b26d14a54b172e11012230c1c78"}
[DEBUG] | WEBSOCKET | 16/08/2024 13:37:42 | GateIO websocket connection: sending message [{"time":1723779462,"id":3,"channel":"spot.book_ticker","event":"subscribe","payload":["BTC_USDT"]}]
[DEBUG] | WEBSOCKET | 16/08/2024 13:37:43 | GateIO websocket connection: message received: {"time":1723779463,"time_ms":1723779463017,"id":3,"conn_id":"1b73153490cc89f8","trace_id":"daa79b26d14a54b172e11012230c1c78","channel":"spot.book_ticker","event":"subscribe","payload":["BTC_USDT"],"result":{"status":"success"},"requestId":"daa79b26d14a54b172e11012230c1c78"}
```

**Review by thrasher- — CHANGES_REQUESTED** [review]
Rather than deviating from the websocket GenerateMessageID helper and rolling out exchange specific counters, I'd opt for the ability to override the default behavior so it still remains unified but configurable

**Review by thrasher- — APPROVED** [review]
Thanks for making those changes 🌞  tACK:
image

**Review by thrasher- — CHANGES_REQUESTED** [review]
Just realised after tACK'ing that this requires test coverage and then should be good to go

**Review by thrasher- — APPROVED** [review]
Changes ACK, thanks!

**Pull request merged**